### PR TITLE
refactor rlhf

### PR DESCRIPTION
**Pull request by hjh0119**
# PR type
- [x] Bug Fix
- [x] New Feature
- [x] Document Updates
- [x] More Models or Datasets Support

# PR information

Refactor RLHF series  algorithms code
- Unified the main function to rlhf.py
- Consolidated Arguments into RLHFArguments. 
- Updated the dependency version of TRL due to KTOTrainer and DPOConfig.

TODO
- [x] Support KTO
- [x] [KTO dataset integration](https://www.modelscope.cn/datasets/AI-ModelScope/ultrafeedback-binarized-preferences-cleaned-kto/summary)
- [x] Support CPO
- [x] Modify the SimPOTrainer to be compatible with trl
- [x] Update doc

Related issue
https://github.com/modelscope/swift/issues/1095
https://github.com/modelscope/swift/issues/1101
https://github.com/modelscope/swift/issues/1138

### CODE REVIEW

File: swift/llm/utils/template.py [hunk 1]
```diff
@@ -228,11 +228,9 @@ def encode(self, example: Dict[str, Any]) -> Tuple[Dict[str, Any], Dict[str, Any
                 'Template is not initialized, please use the `get_template` function to obtain the template.')
         query: Optional[str] = example.get('query', None)
         response: Optional[str] = example.get('response', None)
-        history: Optional[History] = example.get('history', None)
+        history: Optional[History] = example.get('history', [])
```
Comment (Jintao-Huang) [inline swift/llm/utils/template.py]: example = {'history': None, ...} 可能会有风险

File: swift/llm/utils/argument.py [hunk 2]
```diff
@@ -1359,30 +1359,66 @@ def __post_init__(self):
...
-class DPOArguments(SftArguments):
-
+class RLHFArguments(SftArguments):
+    rlhf_type: Literal['dpo', 'orpo', 'simpo', 'kto', 'cpo'] = 'dpo'
     ref_model_type: Optional[str] = field(
         default=None, metadata={'help': f'model_type choices: {list(MODEL_MAPPING.keys())}'})
 
     ref_model_id_or_path: Optional[str] = None
-
+    ref_model_free = False
```
Comment (Jintao-Huang) [inline swift/llm/utils/argument.py]: 少了 : bool

File: swift/llm/utils/argument.py [hunk 3]
```diff
@@ -1359,30 +1359,66 @@ def __post_init__(self):
...
     max_prompt_length: int = 1024
-    beta: float = 0.1
+    beta: Optional[int] = None
     label_smoothing: float = 0.0
-    loss_type: Literal['sigmoid', 'hinge', 'ipo', 'kto_pair'] = 'sigmoid'
+    loss_type: Literal['sigmoid', 'hinge', 'ipo', 'kto_pair', 'simpo'] = 'sigmoid'
     sft_beta: float = 0.1
+    simpo_gamma = 1.0  # reward margin hyperparameter in SimPO
```
Comment (Jintao-Huang) [inline swift/llm/utils/argument.py]: 少了: float maybe

File: swift/llm/utils/argument.py [hunk 4]
```diff
@@ -6,7 +6,7 @@
 import platform
 import sys
 from dataclasses import dataclass, field
-from typing import List, Literal, Optional, Set, Tuple, Union
+from typing import List, Literal, Optional, Set, Tuple, Type, Union
```
Comment (Jintao-Huang) [inline swift/llm/utils/argument.py]: Type是不是没有被使用

File: docs/source/LLM/命令行参数.md [hunk 5]
```diff
@@ -214,14 +214,14 @@ unsloth无新增参数，对已有参数进行调节即可支持：
...
 
-## dpo 参数
-
-dpo参数继承了sft参数, 除此之外增加了以下参数:
-
-- `--ref_model_type`: 对比模型的类型, 可以选择的`model_type`可以查看`MODEL_MAPPING.keys()`.
-- `--ref_model_id_or_path`: 对比模型的本地cache路径, 默认为`None`.
-- `--max_prompt_length`: 最大的提示长度, 该参数会传入DPOTrainer中, 使prompt长度不超过该值的设置, 默认值`1024`.
-- `--beta`: DPO logits的正则项，默认为0.1.
+## RLHF 参数
```
Comment (Jintao-Huang) [inline docs/source/LLM/命令行参数.md:224]: 英文 docs

**Review by tastelikefeet — APPROVED** [review]

**Review by tastelikefeet — APPROVED** [review]

**Pull request merged**